File: CalcPython/EvalVisitor.py
from ExprFuncVisitor import ExprFuncVisitor
from ExprFuncParser import ExprFuncParser
import math

class EvalVisitor(ExprFuncVisitor):
    def __init__(self):
        self.memory = {}
        self.use_degrees = True

    def angle(self, x):
        return math.radians(x) if self.use_degrees else x

    def visitAssign(self, ctx:ExprFuncParser.AssignContext):
        id_ = ctx.ID().getText()
        value = self.visit(ctx.expr())
        self.memory[id_] = value
        return value

    def visitPrintExpr(self, ctx:ExprFuncParser.PrintExprContext):
        value = self.visit(ctx.expr())
        print(value)
        return 0.0

    def visitInt(self, ctx:ExprFuncParser.IntContext):
        return float(ctx.INT().getText())

    def visitFloat(self, ctx:ExprFuncParser.FloatContext):
        return float(ctx.FLOAT().getText())

    def visitId(self, ctx:ExprFuncParser.IdContext):
        id_ = ctx.ID().getText()
        return self.memory.get(id_, 0.0)

    def visitMulDiv(self, ctx:ExprFuncParser.MulDivContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == ExprFuncParser.MUL:
            return left * right
        return left / right

    def visitAddSub(self, ctx:ExprFuncParser.AddSubContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == ExprFuncParser.ADD:
            return left + right
        return left - right

    def visitParens(self, ctx:ExprFuncParser.ParensContext):
        return self.visit(ctx.expr())

    def visitFuncExpr(self, ctx:ExprFuncParser.FuncExprContext):
        value = self.visit(ctx.expr())
        func = ctx.func().getText()
        if func == "sin": return math.sin(self.angle(value))
        if func == "cos": return math.cos(self.angle(value))
        if func == "tan": return math.tan(self.angle(value))
        if func == "sqrt": return math.sqrt(value)
        if func == "log": return math.log10(value)
        if func == "ln": return math.log(value)
        raise Exception(f"Función desconocida: {func}")

    def visitFactorialExpr(self, ctx:ExprFuncParser.FactorialExprContext):
        value = self.visit(ctx.expr())
        if not value.is_integer():
            raise Exception(f"Factorial solo definido para enteros: {value}")
        return self.factorial(int(value))

    def factorial(self, n: int):
        if n < 0:
            raise Exception(f"Factorial no definido para negativos: {n}")
        result = 1
        for i in range(2, n+1):
            result *= i
        return result

File: CalcJava/Java/EvalVisitor.java
import java.util.HashMap;
import java.util.Map;

public class EvalVisitor extends ExprFuncBaseVisitor<Double> {
    /** "memory" para variables */
    Map<String, Double> memory = new HashMap<>();

    /** Configuración: usar grados o radianes para trigonometría */
    private boolean useDegrees = true;

    private double angle(double x) {
        return useDegrees ? Math.toRadians(x) : x;
    }

    /** ID '=' expr NEWLINE */
    @Override
    public Double visitAssign(ExprFuncParser.AssignContext ctx) {
        String id = ctx.ID().getText();
        double value = visit(ctx.expr());
        memory.put(id, value);
        return value;
    }

    /** expr NEWLINE */
    @Override
    public Double visitPrintExpr(ExprFuncParser.PrintExprContext ctx) {
        Double value = visit(ctx.expr());
        System.out.println(value);
        return 0.0;
    }

    /** INT */
    @Override
    public Double visitInt(ExprFuncParser.IntContext ctx) {
        return Double.valueOf(ctx.INT().getText());
    }

    /** FLOAT */
    @Override
    public Double visitFloat(ExprFuncParser.FloatContext ctx) {
        return Double.valueOf(ctx.FLOAT().getText());
    }

    /** ID */
    @Override
    public Double visitId(ExprFuncParser.IdContext ctx) {
        String id = ctx.ID().getText();
        if (memory.containsKey(id)) return memory.get(id);
        return 0.0;
    }

    /** expr op=('*'|'/') expr */
    @Override
    public Double visitMulDiv(ExprFuncParser.MulDivContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));
        return ctx.op.getType() == ExprFuncParser.MUL ? left * right : left / right;
    }

    /** expr op=('+'|'-') expr */
    @Override
    public Double visitAddSub(ExprFuncParser.AddSubContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));
        return ctx.op.getType() == ExprFuncParser.ADD ? left + right : left - right;
    }

    /** '(' expr ')' */
    @Override
    public Double visitParens(ExprFuncParser.ParensContext ctx) {
        return visit(ctx.expr());
    }

    /** func '(' expr ')' */
    @Override
    public Double visitFuncExpr(ExprFuncParser.FuncExprContext ctx) {
        double value = visit(ctx.expr());
        String funcName = ctx.func().getText();

        switch (funcName) {
            case "sin": return Math.sin(angle(value));
            case "cos": return Math.cos(angle(value));
            case "tan": return Math.tan(angle(value));
            case "sqrt": return Math.sqrt(value);
            case "log": return Math.log10(value);
            case "ln": return Math.log(value);
            default: throw new RuntimeException("Función desconocida: " + funcName);
        }
    }

    /** expr '!' */
    @Override
    public Double visitFactorialExpr(ExprFuncParser.FactorialExprContext ctx) {
        double value = visit(ctx.expr());
        long intValue = (long) value;

        if (value != intValue) {
            throw new RuntimeException("Factorial solo definido para enteros: " + value);
        }

        return factorial(intValue);
    }

    private double factorial(long n) {
        if (n < 0) throw new RuntimeException("Factorial no definido para negativos: " + n);
        long result = 1;
        for (long i = 2; i <= n; i++) {
            result *= i;
        }
        return (double) result;
    }
}

